src/librustc_mir/dataflow/mod.rs

```diff
@@ -380,6 +380,62 @@ pub fn state_for_location<'tcx, T: BitDenotation<'tcx>>(loc: Location,
     gen_set.to_dense()
 }
 
+/// Calls `f` with the dataflow state at every location in `mir`.
+/// Ignores blocks that terminate in `unreachable`.
+pub fn for_each_location<'tcx, T: BitDenotation<'tcx>>(
+    mir: &Mir<'tcx>,
+    analysis: &T,
+    result: &DataflowResults<'tcx, T>,
+    mut f: impl FnMut(&HybridBitSet<T::Idx>, Location)
+) {
+    for (block, bb_data) in mir.basic_blocks().iter_enumerated() {
```

Comment:
This seems fragile - I would use at least RPO traversal, with something like `rustc_mir::dataflow::at_location`.
I really wish we had a better cursor API, like `Observer::seek` in `event_flow.rs` in #47954...

Reply:
I don't think it's fragile. We use `on_entry_set_for()` on line 409 to get the exact state at the beginning of each block. No state is shared between blocks. If you still think RPO is necessary, can you explain why?

Reply:
At first, I looked at the `at_location` stuff, but implementing a type that records state and implements a trait just felt like overkill for this application. This function does exactly what I need.

Reply:
Without RPO you might visit dead code - although it turns out that's not related to the `Unreachable` workaround.

As for `at_location` - ugh, I'm really sorry. Can you get in touch with @pnkfelix about the changes to  `rustc_mir::dataflow`?

If you want me to sign off on it, or even regardless of that, please open an issue about designing an ergonomic MIR dataflow API, and put `// FIXME` comments on your addition.